### PR TITLE
feat: DTOSS-3638 modify receive caas file functionapp to read parquet file

### PR DESCRIPTION
**Pull request by sudipnhs**
## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 1]
```diff
@@ -5,86 +5,157 @@ namespace NHS.Screening.ReceiveCaasFile;
...
+                    _logger.LogInformation("Start reading the downloadedfile {name}.", name);
+                    using (var rowReader = ParquetFile.CreateRowReader<ParticipantsParquetMap>(downloadFilePath))
                     {
-                        if (participant != null)
+                        for (var i = 0; i < rowReader.FileMetaData.NumRowGroups; ++i)
                         {
-                            participant.ScreeningId = screeningService.ScreeningId;
-                            participant.ScreeningName = screeningService.ScreeningName;
-                            cohort.Participants.Add(participant);
+                            var values = rowReader.ReadRows(i);
```
Comment (SamRobinson75684) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:80]: is there a more tidy way of mapping the records to the participant object? Could we take this mapping and put it in its own class or example?

Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:84]: done

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 2]
```diff
@@ -5,133 +5,136 @@ namespace NHS.Screening.ReceiveCaasFile;
...
+
+            _logger.LogInformation("Start processing {rowNumber} rows of record from {name} file.", rowNumber, name);
+            SerializeParquetFile(chunks, cohort, name);
+            _logger.LogInformation("All rows processed for file named {name}.", name);
+
         }
         catch (Exception ex)
         {
-            _logger.LogError("{MessageType} validation failed.\nMessage:{ExMessage}\nStack Trace: {ExStackTrace}", ex.GetType().Name, ex.Message, ex.StackTrace);
+            _logger.LogError("Message:{ExMessage}\nStack Trace: {ExStackTrace}", ex.Message, ex.StackTrace);
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:130]: Resolved

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 3]
```diff
@@ -186,4 +186,105 @@ private ScreeningService GetScreeningService(string name)
...
+                var json = JsonSerializer.Serialize(cohort);
+
+                await _callFunction.SendPost(Environment.GetEnvironmentVariable("targetFunction"), json);
+                _logger.LogInformation("Created {CohortCount} Objects.", cohort.Participants.Count);
+            }
+
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError("Message:{ExMessage}\nStack Trace: {ExStackTrace}", ex.Message, ex.StackTrace);
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:219]: Resolved

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 4]
```diff
@@ -5,133 +5,136 @@ namespace NHS.Screening.ReceiveCaasFile;
...
-                    _logger.LogError("File name record count not equal to actual record count. File name count: " + name + "| Actual count: " + rowNumber);
-                    await InsertValidationErrorIntoDatabase(name);
-                    return;
-                }
-                if (rowNumber == 0)
-                {
-                    _logger.LogError("File contains no records. File name:" + name);
-                    await InsertValidationErrorIntoDatabase(name);
-                    return;
+                        if (cohort.Participants.Count == 20000)
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:97]: Resolved

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 5]
```diff
@@ -5,133 +5,136 @@ namespace NHS.Screening.ReceiveCaasFile;
...
+                        participant = await MapParticipant(rec, participant, name, rowNumber);
+
+                        if (participant is null)
                         {
-                            participant.ScreeningId = screeningService.ScreeningId;
-                            participant.ScreeningName = screeningService.ScreeningName;
-                            cohort.Participants.Add(participant);
+                            chunks.Clear();
+                            cohort.Participants.Clear();
+                            _logger.LogError("Invalid data in the file {name}", name);
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:92]: As discussed with @MWClayson-NHS , This is not required.

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 6]
```diff
@@ -5,133 +5,136 @@ namespace NHS.Screening.ReceiveCaasFile;
...
+            {
+                blobStream.CopyTo(fileStream);
+            }
+            var screeningService = GetScreeningService(name);
+            _logger.LogInformation("screeningService {screeningService}", screeningService.ScreeningName);
+
+            _logger.LogInformation("Start reading the downloadedfile {name}.", name);
+            using (var rowReader = ParquetFile.CreateRowReader<ParticipantsParquetMap>(downloadFilePath))
+            {
+                for (var i = 0; i < rowReader.FileMetaData.NumRowGroups; ++i)
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:78]: Comments for RowGroup is added

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 7]
```diff
@@ -186,4 +186,105 @@ private ScreeningService GetScreeningService(string name)
...
+        {
+            _logger.LogError("Message:{ExMessage}\nStack Trace: {ExStackTrace}", ex.Message, ex.StackTrace);
+            await InsertValidationErrorIntoDatabase(filename);
+        }
+    }
+    private async Task<Participant?> MapParticipant(ParticipantsParquetMap rec, Participant participant, string name, int rowNumber)
+    {
+        try
+        {
+            participant.RecordType = Convert.ToString(rec.RecordType);
```
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:227]: As discussed with @MWClayson-NHS , no changes are required. So it's resolved

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 8]
```diff
@@ -5,133 +5,138 @@ namespace NHS.Screening.ReceiveCaasFile;
...
+
+            _logger.LogInformation("Start processing {rowNumber} rows of record from {name} file.", rowNumber, name);
+            SerializeParquetFile(chunks, cohort, name);
+            _logger.LogInformation("All rows processed for file named {name}.", name);
+
         }
         catch (Exception ex)
         {
-            _logger.LogError("{MessageType} validation failed.\nMessage:{ExMessage}\nStack Trace: {ExStackTrace}", ex.GetType().Name, ex.Message, ex.StackTrace);
+            _logger.LogError("Stack Trace: {ExStackTrace}\nMessage:{ExMessage}", ex.StackTrace, ex.Message);
```
Comment (MWClayson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:132]: ` _logger.LogError(ex,"Stack Trace: {ExStackTrace}\nMessage:{ExMessage}", ex.StackTrace, ex.Message);`

the above is what I mean when i say put the stack trace first in the log statements.

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj [hunk 9]
```diff
@@ -15,6 +16,8 @@
         <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.16.4" />
         <PackageReference Include="Microsoft.ApplicationInsights.WorkerService" Version="2.21.0" />
         <PackageReference Include="Microsoft.Azure.Functions.Worker.ApplicationInsights" Version="1.1.0" />
+        <PackageReference Include="Parquet.Net" Version="4.25.0-pre.2" />
```
Comment (MWClayson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj:19]: Are both packages Parquet.net and ParquetSharp required?

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj [hunk 10]
```diff
@@ -7,6 +7,7 @@
         <Nullable>enable</Nullable>
     </PropertyGroup>
     <ItemGroup>
+        <PackageReference Include="AutoMapper" Version="13.0.1" />
```
Comment (MWClayson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj:10]: Not sure if this is needer either

Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj:10]: Yes, you are right. I just forgot to remove the reference.

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj
[the remark this answers is not in the document]
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj:1]: Yes, you are right. Only ParquetSharp Package is in use. I forgot to remove Parquet.Net reference.
Parquet.Net references are now removed

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj (hunk 9 above)
Comment (sudipnhs) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.csproj]: Yes we are right. "Parquet.Net" is not in use, so removed it's references from the project.

**Comment by MWClayson-NHS** [comment]
Please can you remerge main back in to this branch and then I'll approve

**Pull request closed without merging**